README
File-Vctools Version 0.09

NAME

  Compare different versions of text files and identify changes

INSTALLATION

  To install this module, run the following commands:

  perl Makefile.PL
  make
  make test
  make install

SYNOPSIS

  File::Vctools is a collection of utility programs that help you
  to organise your projects. Currently, File::Vctools only runs on
  Windows and on Linux, but if somebody wants to run it on a platform
  other than Windows or Linux, the conversion should be easy. If you
  already have converted File::Vctools to a different platform, let me
  know at <klaus03@gmail.com>, I appreciate any feedback.

  Here is a first program that runs the components of File::Vctools:

  use strict;
  use warnings;
  use 5.010;

  use File::Vctools qw(get_mpath get_difftool);
  use File::Spec;
  use File::Slurp;
  use File::Temp qw(tempdir);

  my $difftool = get_difftool();
  my $mpath    = get_mpath();
  my $cwd      = File::Spec->rel2abs('.');
  END { chdir $cwd if defined $cwd; }
  my $tempdir  = tempdir(CLEANUP => 1);

  say "difftool is $difftool";
  say "mpath    is $mpath";
  say "cwd      is $cwd";
  say "tempdir  is $tempdir";
  say '';

  mkdir File::Spec->catdir($tempdir, 'XmlRepo');
  mkdir File::Spec->catdir($tempdir, 'test_arch');
  mkdir File::Spec->catdir($tempdir, 'Original');
  mkdir File::Spec->catdir($tempdir, 'P001');

  $ENV{VCTOOLDIR} = File::Spec->catdir($tempdir, 'XmlRepo');

  write_file(File::Spec->catdir($tempdir, 'XmlRepo', 'vc_parameter.xml'),
    qq{<?xml version="1.0" encoding="iso-8859-1"?>\n},
    qq{<vc>\n},
    qq{  <archive path="}, File::Spec->catdir($tempdir, 'test_arch'),
      qq{" />\n},
    qq{</vc>\n},
  );

  write_file(File::Spec->catfile($tempdir, 'Original', 'file.txt'),
    qq{Line001\n},
    qq{Line002\n},
    qq{Line003\n},
    qq{Line004\n},
  );

  chdir File::Spec->catdir($tempdir, 'P001') or die "Error-0010: chdir $!";

  system $^X, File::Spec->catfile($mpath, 'vc_init.pl');
  system $^X, File::Spec->catfile($mpath, 'vc_checkout.pl'),
    File::Spec->catfile($tempdir, 'Original', 'file.txt');

  write_file(File::Spec->catfile('Work', 'F_file_Z001.txt'),
    qq{Line001\n},
    qq{Line002 ***\n},
    qq{Line003\n},
    qq{Line004\n},
  );

  system $^X, File::Spec->catfile($mpath, 'vc_apply.pl');
  system $^X, File::Spec->catfile($mpath, 'vc_status.pl'), '-a', '-o';
  system $^X, File::Spec->catfile($mpath, 'vc_list.pl');
  system $^X, File::Spec->catfile($mpath, 'vc_reset.pl');

  chdir 'Data' or die "Error-0020: chdir $!";

  write_file('orig.txt',
    qq{Orig001\n},
    qq{Orig002\n},
    qq{Orig003\n},
    qq{Orig004\n},
    qq{Orig005\n},
  );

  write_file('patch.txt',
    qq{\@\@ -2,4 +2,4 \@\@\n},
    qq{ Orig002\n},
    qq{-Orig003\n},
    qq{+Orig003***\n},
    qq{ Orig004\n},
  );

  system $^X, File::Spec->catfile($mpath, 'vc_merge.pl'),
    '--input=orig.txt', '--diff=patch.txt', '--output=out.txt';

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc File::Vctools

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=File-Vctools

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/File-Vctools

    CPAN Ratings
        http://cpanratings.perl.org/d/File-Vctools

    Search CPAN
        http://search.cpan.org/dist/File-Vctools/


COPYRIGHT AND LICENSE

Copyright (C) 2010 by Klaus Eichner

All rights reserved. This program is free software; you can redistribute
it and/or modify it under the terms of the artistic license,
see http://www.opensource.org/licenses/artistic-license-1.0.php


Changes
Revision history for File-Vctools

0.01    30 Mar 2010
        First version, released on an unsuspecting world.

0.02    10 Apr 2010
        Change header in documentation File-Vctools_de.pod and File-Vctools_fr.pod

0.03    20 Apr 2010
        Complete documentation also in French and in German
        I added the idea of a "current project" to File::Vctools,
        as a consequence, vc_list.pl now shows an additional line to indicate the current project
        and vc_apply.pl aborts if you try to apply when your project is not current
        ==> internally I had to restructure "D_coutlist.dat" to introduce an additional level of hash-keys
        ==> {D_pmtab => ...} now contains what used to be under root
        ==> and there is an additional { D_pmdef => {project} } to hold the name of the current project
        ==> vc_status.pl now has a char-by-char comparison
        ==> minor changes in vc_merge.pl

0.04    18 Jul 2010
        The command 'vc_checkout.pl' now checks and eliminates all files that
          are not listed in 'B_Flist.xml'.
        Also, call XML::Reader version 0.37 the new way:
        XML::Reader->new($cnst_VcParam, {mode => 'branches'}
        in 'vc_checkout.pl': when you encounter a situation
          "Error-0050: Found directory file '$_' which does not exist in archive"
          then warn and carry on instead of just dying.
        in 'r_renew.pl' (that's set up by 'vc_init.pl') add a second call to
          'vc_checkout.pl'. 'r_renew.pl' (again set up by 'vc_init.pl') now looks
          as follows:
            go('vc_reset');
            go('vc_checkout', '-z');
            go('vc_list', '-n', '-z', '-c');
            go('vc_checkout', '-z');
            go('vc_apply', '-q');

0.05    22 Jul 2010
        Refactoring 'vc_checkout.pl': compare all files in /Work/ against the
          checkout-status.
        As a consequence, 'vc_checkout.pl' has grown up and does not rely anymore
          on 'vc_list.pl' being run first, therefore the call to 'vc_list.pl' could
          be removed from 'r_renew.pl'.
        'r_renew.pl' is now much simpler, it looks as follows:
            go('vc_reset');
            go('vc_checkout', '-z');
            go('vc_apply', '-q');

0.06    24 Jul 2010
        add housekeeping to 'vc-checkout.pl' (clean up after a file has been removed from
        'B-Flist.xml')
        reformat and renumber testcases in 't/0010-test.t'

0.07    04 Aug 2010
        Add "EXE_FILES => [ 'bin/vc_apply.pl', 'bin/vc_checkout.pl', ... ] to Makefile.PL

0.08    24 Aug 2010
        Fix bug in vc_checkout.pl where two identical filenames in different directories
        resulted in a single checkout file, one file content overwrote the other.

0.09    25 Aug 2010
        The fix in version 0.08 had a minor problem: the internal counter to produce
        a new checkout file was incremented too often. To be correct, the internal counter
        in vc_checkout.pl is incremented only if the current checkout file is permanent.


Makefile.PL
use 5.010;

use ExtUtils::MakeMaker;

unless ($^O eq 'MSWin32' or $^O eq 'linux') {
    die "Error-0010: File::Vctools can only run on \$^O = 'MSWin32' or 'linux', but found \$^O = '$^O'";
}

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'File::Vctools',
    VERSION_FROM      => 'lib/File/Vctools.pm', # finds $VERSION
    PREREQ_PM         => { 
                           'XML::Reader'     => 0.37,
                           'Algorithm::Diff' => 1.19_02,
                           'File::Slurp'     => 9999.12,
                         }, # e.g., Module::Name => 1.1
    EXE_FILES         => [ 'bin/vc_apply.pl', 'bin/vc_checkout.pl', 'bin/vc_init.pl',
                           'bin/vc_list.pl',  'bin/vc_merge.pl',    'bin/vc_reset.pl',
                           'bin/vc_status.pl'],
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/File/Vctools.pm', # retrieve abstract from module
       AUTHOR         => 'Klaus Eichner <klaus03@gmail.com>') : ()),
);
